### ISSUE TITLE
ClassDeclaration/ClassExpression option for the indent rule

### ISSUE THREAD

**Issue by nemoxps** [eslint/eslint#8230]
**Tell us about your environment**

* **ESLint Version:** v3.17.1 (using the [`indent-rewrite`](https://github.com/eslint/eslint/pull/7618) branch)
* **Node Version:** v7.7.2
* **npm Version:** v4.1.2

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
```js
module.exports = {
    root: true,
    parserOptions: {
        ecmaVersion: 8
    },
    rules: {
        indent: [2, 2, {
            SwitchCase: 2,
            VariableDeclarator: 2,
            outerIIFEBody: 2,
            MemberExpression: 2,
            FunctionDeclaration: {
                parameters: 2,
                body: 2
            },
            FunctionExpression: {
                parameters: 2,
                body: 2
            },
            CallExpression: {
                arguments: 2
            },
            ArrayExpression: 2,
            ObjectExpression: 2
        }]
    }
};
```

**What did you do? Please include the actual source code causing the issue.**
```js
class a {
    m() {
        "hmm";
    }
}
let b = class {
    m() {
        "hmm";
    }
};
```

**What did you expect to happen?**
There should be an option to set the indentation for `ClassDeclaration` and `ClassExpression`.

**What actually happened? Please include the actual, raw output from ESLint.**
```
c:\nodejs\test\test.js
  2:5  error  Expected indentation of 2 spaces but found 4  indent
  3:9  error  Expected indentation of 6 spaces but found 8  indent
  4:5  error  Expected indentation of 2 spaces but found 4  indent
  7:5  error  Expected indentation of 2 spaces but found 4  indent
  8:9  error  Expected indentation of 6 spaces but found 8  indent
  9:5  error  Expected indentation of 2 spaces but found 4  indent

✖ 6 problems (6 errors, 0 warnings)
```

`$ eslint test.js --fix`
Lints to:
```js
class a {
  m() {
      "hmm";
  }
}
let b = class {
  m() {
      "hmm";
  }
};
```

**Comment by platinumazure** [comment]
Hi @nemoxps, thanks for the issue.

It looks like you're reporting a bug, but in reality we never supported an option for ClassDeclaration/ClassExpression indentation. Would you mind instead filling out our [rule change proposal template](https://github.com/eslint/eslint/blob/master/templates/rule-change-proposal.md) please? Thanks!

**Comment by eslintbot** [comment]
Hi @nemoxps, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by nemoxps** [comment]
## What rule do you want to change?
indent

## Does this change cause the rule to produce more or fewer warnings?
Same. Or more, if you set the options.

## How will the change be implemented? (New option, new default behavior, etc.)?
New options.
```js
ClassDeclaration: {
    type: "integer",
    minimum: 0
},
ClassExpression: {
    type: "integer",
    minimum: 0
},
```

## Please provide some example code that this change will affect:

```js
class a {
m() {
}
}
let b = class {
m() {
}
};
```

## What does the rule currently do for this code?
Uses default indentation.
```js
class a {
  m() {
  }
}
let b = class {
  m() {
  }
};
```

## What will the rule do after it's changed?
```js
indent: [2, 2, {
    ClassDeclaration: 2,
    ClassExpression: 2,
}]
```
```js
class a {
    m() {
    }
}
let b = class {
    m() {
    }
};
```

```js
indent: [2, 2, {
    ClassDeclaration: 3,
    ClassExpression: 3,
}]
```
```js
class a {
      m() {
      }
}
let b = class {
      m() {
      }
};
```

**Comment by kaicataldo** [comment]
This option feels weird to me but I'm not strongly against it.

**Comment by nemoxps** [comment]
@kaicataldo Do you mind to elaborate why you think this option feels weird? I mean, there is an option for functions, why shouldn't there be an option for classes?

**Comment by not-an-aardvark** [comment]
Thanks for the proposal.

We're trying to limit the number of options we add to the `indent` rule, since it already has a very large number of options and it's difficult to maintain. So I suppose my question would be whether this is a common enough preferences to be worth the complexity of adding it.

**Comment by nemoxps** [comment]
@not-an-aardvark I just wanted to mention that the added complexity is **just 15 lines** of code (in case of I did it right, at least the tests I've written say so).

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.

**Comment by mqklin** [comment]
@nemoxps have you written a rule for this?  I need this rule too, because eslint forces me to write everywhere 2 indent, except classes. So weird

**Comment by mqklin** [comment]
@kaicataldo I think it's weird not to include this option, because we have options for all expressions (Array, Object, Functions) except Classes.

**Comment by nemoxps** [comment]
@mqklin Sorry for being a bit late. Yes, I wrote my own indent patch. You may check out [my plugin](https://github.com/nemoxps/eslint-plugin-nemoxps). I just updated the indent rule today, so it has all the latest eslint-indent patches.

**Installation**
```sh
npm install nemoxps/eslint-plugin-nemoxps
```

**Add it to your ***package.json*****
```json
{
    ...
    "peerDependencies": {
        "eslint-plugin-nemoxps": "nemoxps/eslint-plugin-nemoxps#semver:^0.4.0"
    }
}
```

**Add it to your ***.eslintrc.js*****
```js
module.exports = {
    plugins: ['nemoxps'],
    rules: {
        // TODO: Enable when ClassDeclaration/ClassExpression options are implemented.
        indent: 0,
        'nemoxps/indent': [2, 2, {
            // Your original indent config
            ClassDeclaration: 2,
            ClassExpression: 2,
        }],
    },
};
```

Though, I'd rather like to see this in ESLint natively. It's just a few lines...

**Comment by mqklin** [comment]
@nemoxps I've just installed this plugin and I have this error: `Error while running ESLint: Cannot find module '../../eslint/node_modules/lodash'.`

**Comment by nemoxps** [comment]
@mqklin That error is caused by this line: `const lodash = require('../../eslint/node_modules/lodash');`
To explain this line: I'm assuming that ESLint and my plugin both are installed globally. It is in my case and it just works fine.
If you need it to work locally, just fill an issue and I'm going to fix it.

**Comment by not-an-aardvark** [comment]
It would be better to depend on `lodash`, `ast-utils`, and `functional-red-black-tree` directly in your plugin rather than loading ESLint's versions of the modules at runtime. That way, your plugin won't break if ESLint modifies its dependencies or helper functions in a future release.
